### PR TITLE
Ensure no extension method applies to System.Void

### PR DESCRIPTION
**Pull request by genlu**
Fix https://github.com/dotnet/roslyn/issues/65020

This is not an issue in VB because there's already a single check for regular members and extension methods exist https://sourceroslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Binding/Binder_Lookup.vb,1928

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs [hunk 1]
```diff
@@ -745,7 +745,7 @@ public MethodSymbol ReduceExtensionMethod(TypeSymbol receiverType, CSharpCompila
                 throw new ArgumentNullException(nameof(receiverType));
             }
 
-            if (!this.IsExtensionMethod || this.MethodKind == MethodKind.ReducedExtension)
+            if (!this.IsExtensionMethod || this.MethodKind == MethodKind.ReducedExtension || receiverType.SpecialType == SpecialType.System_Void)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs:748]: This doesn't look like a change in SemanticModel. Where are instance members getting dropped?

Comment (genlu) [reply src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs:748]: That check is here
https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Binder_Lookup.cs,1984

`SemanticModel.LookUpSymbols` first calls (binder.AddMemberLookupSymbolsInfo` to get regular members and then calls `MethodSymbol.ReduceExtensionMethod` for reduced extension methods
https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Compilation/CSharpSemanticModel.cs,1633

File: src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs [hunk 2]
```diff
@@ -4079,5 +4079,29 @@ void Validator(ModuleSymbol module)
...
+            var compilation = CreateCompilationWithMscorlib40AndSystemCore(source);
+            compilation.VerifyDiagnostics();
+
+            var extensionMethod = compilation.GlobalNamespace.GetMember<NamedTypeSymbol>("C").GetMember<MethodSymbol>("M");
+            Assert.True(extensionMethod.IsExtensionMethod);
+
+            var systemVoidType = compilation.GetSpecialType(SpecialType.System_Void);
+            Assert.Equal(SpecialType.System_Void, systemVoidType.SpecialType);
+
+            var reduced = extensionMethod.ReduceExtensionMethod(systemVoidType, null!);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs:4103]: >null! [](http://example.com/codeflow?start=80&length=5)

Why not pass compilation?

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs [hunk 3]
```diff
@@ -745,7 +745,7 @@ public MethodSymbol ReduceExtensionMethod(TypeSymbol receiverType, CSharpCompila
                 throw new ArgumentNullException(nameof(receiverType));
             }
 
-            if (!this.IsExtensionMethod || this.MethodKind == MethodKind.ReducedExtension)
+            if (!this.IsExtensionMethod || this.MethodKind == MethodKind.ReducedExtension || receiverType.IsVoidType())
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs:748]: >|| receiverType.IsVoidType() [](http://example.com/codeflow?start=90&length=28)

Do we need to make similar change for VB?

File: src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs [hunk 4]
```diff
@@ -4079,5 +4079,29 @@ void Validator(ModuleSymbol module)
 
             CompileAndVerify(source, validator: Validator, options: TestOptions.ReleaseDll);
         }
+
+        [Fact]
+        [WorkItem(65020, "https://github.com/dotnet/roslyn/issues/65020")]
+        public void ReduceExtensionsMethodOnReceiverTypeSystemVoid()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs:4085]: >ReduceExtensionsMethodOnReceiverTypeSystemVoid [](http://example.com/codeflow?start=20&length=46)

Let's add similar test for VB

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs (hunk 3 above)
Comment (genlu) [reply src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs:748]: I don't think so. VB has the check [here](https://sourceroslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Binding/Binder_Lookup.vb,1928), which also guards extension method look-up.

File: src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs (hunk 4 above)
Comment (genlu) [reply src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs:4085]: Done

File: src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs (hunk 2 above)
Comment (genlu) [reply src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs:4103]: Added an assertion with compilation passed in

File: src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs:4085]: It looks like the added test is not for the ```ReduceExtensionMethod``` API. The goal to have the test is to ensure that the API behaves consistently for both languages

File: src/Compilers/VisualBasic/Test/Semantic/Compilation/SemanticModelLookupSymbolsAPITests.vb [hunk 5]
```diff
@@ -9,6 +9,7 @@ Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
 Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
 
 Imports Roslyn.Test.Utilities
+Imports Roslyn.Test.Utilities.TestMetadata
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/Compilation/SemanticModelLookupSymbolsAPITests.vb:12]: >Roslyn.Test.Utilities.TestMetadata [](http://example.com/codeflow?start=8&length=34)

Why is this needed?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 3)

File: src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs (hunk 4 above)
Comment (genlu) [reply src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs:4085]: Ah OK. I added a semantic model test because the [check for System.Void](https://sourceroslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Binding/Binder_Lookup.vb,1928) happens before the [call to ReduceExtensionMethod](https://sourceroslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Binding/Binder_Lookup.vb,1220) I have changed the test to cover ReduceExtensionMethod directly

File: src/Compilers/VisualBasic/Test/Semantic/Compilation/SemanticModelLookupSymbolsAPITests.vb (hunk 5 above)
Comment (genlu) [reply src/Compilers/VisualBasic/Test/Semantic/Compilation/SemanticModelLookupSymbolsAPITests.vb]: removed

**Review by 333fred — APPROVED** [review]

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/ExtensionMethods/ExtensionMethodTests.vb [hunk 6]
```diff
@@ -2535,6 +2535,36 @@ End Module
...
+End Structure
+Module E
+    <Extension()>
+    Friend Sub ExtMethod(o As ValueType)
+    End Sub
+End Module
+]]></file>
+</compilation>, {Net40.SystemCore})
+
+            Dim tree = compilation.SyntaxTrees.Single()
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/ExtensionMethods/ExtensionMethodTests.vb:2556]: >Dim tree = compilation.SyntaxTrees.Single() [](http://example.com/codeflow?start=12&length=43)

It looks like this local isn't used

**Comment by AlekseyTs** [comment]
@genlu Please avoid force pushing changes into PRs under review. This often confuses review tools, and prevents us from reviewing changes in an incremental fashion.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 4), modulo the unused local in VB test

**Comment by AlekseyTs** [comment]
@genlu Please squash commits during merging
